### PR TITLE
Catalyst and PL Control unification

### PR DESCRIPTION
**Pull request by maliasadi**
**Context:**
Restructure `catalyst.ctrl` to inherit from the PL Controlled class.

This provides feature parity for single operations. For quantum functions, however, this support is not fully compatible with PL because PL only instantiates the `Controlled` and `ControlledOp` classes of a single base operation and general operator and returns a list of `Controlled` instances while Catalyst intends to return a single controlled instance with a list of ops nested inside of it.

**Description of the Change:**
- Rename the current `QCtrl(HybirdOp)` class to `HybridCtrl(HybridOp)`. This class provides support for computing the control of single-op and quantum functions (multiple-ops).
- Create new classes `catalyst.ControlledOp`, `catalyst.ControlledBase` derived from `catalyst.Controlled` to provide feature parity with PL for single-operation and single-operator. The tracer was also updated to make this class QJIT-compatible.
- Create a new wrapper class `CtrlCallable` to produce a QJIT-compatible ctrl instance.
- Update `catalyst.ctrl` to create a `CtrlCallable` instance based on the input base operation/callable.
- Add PL's Controlled unit tests to check the feature parity of `catalyst.ctrl`.

**Benefits:**
- Better feature parity of `catalyst.ctrl` with  PL's ctrl for single ops.
- Allow `catalyst.ctrl` to be used outside the QJIT or QNode context similar to the PL ctrl method.

**Possible Drawbacks:**

**Related Tickets:**
[sc-60924]
[sc-60925]

### CODE REVIEW

File: frontend/test/pytest/test_autograph.py [hunk 1]
```diff
@@ -72,7 +72,7 @@ def val(self):
         return self.ref
 
 
-class TestSourceCodeInfo:
+class TestSourceCodeXXXInfo:
```
Comment (erick-xanadu) [inline frontend/test/pytest/test_autograph.py]: Why the name change here?

File: frontend/test/pytest/test_autograph.py [hunk 2]
```diff
@@ -72,7 +72,7 @@ def val(self):
         return self.ref
 
 
-class TestSourceCodeXXXInfo:
+class TestSourceCodeInfo:
```
Comment (erick-xanadu) [inline frontend/test/pytest/test_autograph.py]: Ah, never mind, changed back here.

File: frontend/catalyst/api_extensions/quantum_operators.py [hunk 3]
```diff
@@ -338,15 +334,66 @@ def cond_fn():
         (len(control) if isinstance(control, Sized) else 1)
         != (len(control_values) if isinstance(control_values, Sized) else 1)
     ):
-        raise ValueError(
-            f"Length of the control_values ({len(control_values)}) must be None or equal "
-            f"to the lenght of control ({len(control)})"
-        )
+        raise ValueError("control_values should be the same length")
```
Comment (erick-xanadu) [inline frontend/catalyst/api_extensions/quantum_operators.py:337]: I liked the previous error message more. It had more information. Can it be changed back?

File: frontend/test/pytest/test_quantum_control.py [hunk 4]
```diff
@@ -494,98 +524,921 @@ def use_legacy_opmath():
...
-        assert op.base is base
-        assert op.hyperparameters["base"] is base
-        assert op.name == "C(PauliX)"
+        assert op.work_wires == Wires(("aux"))
+
+        assert op.name == "C(TempOperator)"
 
         assert op.num_params == 0
-        assert op.parameters == []
+        assert op.parameters == []  # pylint: disable=use-implicit-booleaness-not-comparison
```
Comment (erick-xanadu) [inline frontend/test/pytest/test_quantum_control.py:561]:
```suggestion
        assert not []
```

File: frontend/catalyst/jax_tracer.py [hunk 5]
```diff
@@ -416,9 +416,9 @@ def trace_quantum_tape(
     def _bind_native_controlled_op(qrp, op, controlled_wires, controlled_values):
         # For named-controlled operations (e.g. CNOT, CY, CZ) - bind directly by name. For
         # `Controlled(OP)` bind OP with native quantum control syntax.
-        if op.__class__ in {Controlled, ControlledOp, ControlledQubitUnitary} or isinstance(
+        if isinstance(
             op, catalyst.api_extensions.quantum_operators.HybridControlledOp
-        ):
+        ) or op.__class__ in {Controlled, ControlledOp, ControlledQubitUnitary}:
```
Comment (erick-xanadu) [inline frontend/catalyst/jax_tracer.py]:
```suggestion
        classinfo = catalyst.api_extensions.quantum_operators.HybridControlledOp
        is_catalyst_ctrl = isinstance(op, classinfo)
        pl_ctrl_classes = { Controlled, ControlledOp, ControlledQubitUnitary }
        is_pl_ctrl = op.__class__ in pl_ctrl_classes
        is_ctrl = is_catalyst_ctrl or is_pl_ctrl
        if is_ctrl:
```

Why can't we use `isinstance(op, pl_ctrl_classes)` here?

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 3 above)
Comment (maliasadi) [reply frontend/catalyst/api_extensions/quantum_operators.py:337]: yeah, I agree.. I'll change it back!

File: frontend/test/pytest/test_quantum_control.py (hunk 4 above)
Comment (maliasadi) [reply frontend/test/pytest/test_quantum_control.py:561]: I don't understand this suggestion. :D

File: frontend/catalyst/jax_tracer.py (hunk 5 above)
Comment (maliasadi) [reply frontend/catalyst/jax_tracer.py]: isn't the current implementation shorter and clear? I don't mind to accept your suggestion but just asking.

File: frontend/test/pytest/test_quantum_control.py (hunk 4 above)
Comment (erick-xanadu) [reply frontend/test/pytest/test_quantum_control.py:561]: Sorry about it, I just connected it above.

File: frontend/catalyst/jax_tracer.py (hunk 5 above)
Comment (erick-xanadu) [reply frontend/catalyst/jax_tracer.py]: I prefer my if conditions to be simple and read similar to English. I am not sure you saw my question at the end, why use `op.__class__` vs `isinstance`?

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 3 above)
Comment (maliasadi) [reply frontend/catalyst/api_extensions/quantum_operators.py:337]: By the way I had to change the error message here to make it consistent with PL.

File: frontend/catalyst/jax_tracer.py (hunk 5 above)
Comment (maliasadi) [reply frontend/catalyst/jax_tracer.py]: Ah, just saw your question!  `ctrl(callables)` is now an instance of `HybridControlledOp` which needs to be treated as a valid PL controlled class by tracer. Removing this check causes the tracer to capture the ctrl op as a quantum gate and generate an incorrect JAXPR.

File: frontend/catalyst/api_extensions/quantum_operators.py [hunk 6]
```diff
@@ -343,47 +339,8 @@ def cond_fn():
...
-        QueuingManager.remove(f)
-
-        def _callee():
-            QueuingManager.append(f)
-
-        return _call_handler(_callee=_callee)
-
-    else:
-        raise ValueError(f"Expected a callable or a qml.Operator, not {f}")  # pragma: no cover
+    res = QCtrlCallable(op, control, control_values=control_values, work_wires=work_wires)
```
Comment (dime10) [inline frontend/catalyst/api_extensions/quantum_operators.py:342]: Nice, I see you already renamed the other classes to better match the PL names (Controlled and HybridControlled), could we also eliminate all other references to "QCtrl"? This class could just CtrlCallable or ControlCallable.
To be fair HybridControlled does sound a bit weird, maybe HybridControl would be sufficient?

Comment (erick-xanadu) [reply frontend/catalyst/api_extensions/quantum_operators.py:342]: Does Hybrid make sense here though? Are we using classical instructions somehow? Naming is hard :sweat_smile:

File: frontend/catalyst/jax_tracer.py (hunk 5 above)
Comment (erick-xanadu) [reply frontend/catalyst/jax_tracer.py]: > Ah, just saw your question! ctrl(callables) is now an instance of HybridControlledOp which needs to be treated as a valid PL controlled class by tracer. Removing this check causes the tracer to capture the ctrl op as a quantum gate and generate an incorrect JAXPR.

I am not sure I still understand the differences between isinstance and checking `.__class__` entirely. Just to be explicit, would changing:

```python
        pl_ctrl_classes = { Controlled, ControlledOp, ControlledQubitUnitary }
        is_pl_ctrl = op.__class__ in pl_ctrl_classes
```

to

```python
        pl_ctrl_classes = { Controlled, ControlledOp, ControlledQubitUnitary }
        is_pl_ctrl = isinstance(op, pl_ctrl_classes)
```

Not be the same? I know you didn't introduce this change and you are only preserving the old behaviour, but it is an interesting way to check what I think we want to check.

**Review by erick-xanadu — APPROVED** [review]
I'll let you coordinate with David too since he has more context than I do regarding this ticket. Cheers! :partying_face:

File: frontend/catalyst/jax_tracer.py (hunk 5 above)
Comment (maliasadi) [reply frontend/catalyst/jax_tracer.py]: @erick-xanadu these are not the same:
- `op.__class__` checks if the exact type of `op` is in `pl_ctrl_classes`, while
-  `isinstance(op, pl_ctrl_classes)` checks if `op` is an instance of these classes or any of their subclasses.

For example, `qml.CNOT` is an instance of `pennylane.ops.op_math.controlled_ops.CNOT` which is a subclass of `Controlled`. Using the latter instead of `op.__class__` would result the decomposition of `CNOT` to `PauliX`! This is not ideal as this gate is supported by the runtime. Thus, `op.__class__` is used here to ensure controlled based operations, those are supported by Runtime C-API, are not decomposed.

Comment (maliasadi) [reply frontend/catalyst/jax_tracer.py]: @erick-xanadu these are not the same:
- `op.__class__` checks if the exact type of `op` is in `pl_ctrl_classes`, while
-  `isinstance(op, pl_ctrl_classes)` checks if `op` is an instance of these classes or any of their subclasses.

For example, `qml.CNOT` is an instance of `pennylane.ops.op_math.controlled_ops.CNOT` which is a subclass of `Controlled`. Using the latter instead of `op.__class__` would result the decomposition of `CNOT` to `PauliX`! This is not ideal as this gate is supported by the runtime. Thus, `op.__class__` is used here to ensure controlled based operations, those are supported by Runtime C-API, are not decomposed, I think.

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 6 above)
Comment (maliasadi) [reply frontend/catalyst/api_extensions/quantum_operators.py:342]: I use `CtrlCallable` and it eliminates the ambiguity between Controlled and Control ;)

File: frontend/catalyst/jax_tracer.py (hunk 5 above)
Comment (erick-xanadu) [reply frontend/catalyst/jax_tracer.py]: Awesome! Thank you :)

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 6 above)
Comment (dime10) [reply frontend/catalyst/api_extensions/quantum_operators.py]: We are in fact allowing classical instructions in the body of the control operator :) The main differentiator to a quantum only version which could hold a nested tape would be that this op contains logic to handle a nested jaxpr scope. 
Naming is definitely hard 😅

File: frontend/catalyst/jax_tracer.py [hunk 7]
```diff
@@ -476,7 +481,9 @@ def _bind_native_controlled_op(qrp, op, controlled_wires, controlled_values):
 
     for op in ops:
         qrp2 = None
-        if isinstance(op, HybridOp):
+        if isinstance(op, HybridOp) and not isinstance(
```
Comment (dime10) [inline frontend/catalyst/jax_tracer.py:484]:
```suggestion
        # We need to exclude HybridCtrl here because single-op control instances are kept
        # as instances of the Catalyst Controlled class, which also inherits from HybridCtrl,
        # but should be translated to JAXPR as a regular PennyLane Controlled op.
        # Native HybridCtrl operations are not yet supported in the compiler.
        if isinstance(op, HybridOp) and not isinstance(
```

An alternative to this problem is to not make catalyst Controlled inherit from HybridCtrl (or even dispatching to PennyLane Controlled directly in the single op case), but that does set us up to support control meta ops in the compiler, and keeps it more inline with adjoint.

File: frontend/test/lit/test_quantum_control.py [hunk 8]
```diff
@@ -130,9 +130,7 @@ def test_native_controlled_unitary():
     @qml.qnode(dev)
     # CHECK-LABEL: public @jit_native_controlled_unitary
     def native_controlled_unitary():
-        # CHECK: [[out:%.+]], [[out_ctrl:%.+]]:3 = quantum.unitary
-        # CHECK-SAME: ctrls
-        # CHECK-SAME: ctrlvals(%true, %true, %true)
+        # CHECK: [[out:%.+]] = quantum.unitary
         qml.ctrl(
```
Comment (dime10) [inline frontend/test/lit/test_quantum_control.py:134]: Why this change?

File: frontend/catalyst/api_extensions/quantum_operators.py [hunk 9]
```diff
@@ -676,7 +774,7 @@ def qctrl_distribute(
     ops2 = []
     for op in tape.operations:
```
Comment (dime10) [inline frontend/catalyst/api_extensions/quantum_operators.py:775]: One more change we could carry over from the adjoint impl is swapping `create_controlled_op` further down in this function to `qml.ctrl`.

The former always creates a PL Controlled instance, while the latter will create either a PL or Catalyst Controlled instance depending on context. Since the Catalyst version is now fully compatible the need for the existence of `create_controlled_op` is probably eliminated entirely.

**Review by dime10 — CHANGES_REQUESTED** [review]
Great work, only a few comments but otherwise we are good to go 💯

File: frontend/catalyst/api_extensions/quantum_operators.py [hunk 10]
```diff
@@ -238,15 +239,12 @@ def loop_fn(i):
...
-        adj = adj()
-
-    return adj
+    # Return an instantiated version of the Adjoint class if we receive an operator instance.
+    return adj() if isinstance(f, Operator) else adj
 
 
 def ctrl(
-    f: Union[Callable, Operator],
+    op: Union[Callable, Operator],
```
[the remark this answers is not in the document]
Comment (maliasadi) [reply frontend/catalyst/api_extensions/quantum_operators.py:247]: Just followed the name convention in PL, although `op` in Catalyst can be a general Callable while in PL only Operators and functions are allowed: https://github.com/PennyLaneAI/pennylane/blob/7f97ea735b3c43b43a0ffe34040bffaa49603268/pennylane/ops/op_math/controlled.py#L38 They use `op` with the following docstring:
```
        op (function or :class:`~.operation.Operator`): A single operator or a function that applies pennylane operators.
```

File: frontend/catalyst/jax_primitives.py [hunk 11]
```diff
@@ -77,7 +77,7 @@
 from catalyst.utils.extra_bindings import FromElementsOp, TensorExtractOp
 from catalyst.utils.types import convert_shaped_arrays_to_tensors
 
-# pylint: disable=unused-argument,abstract-method,too-many-lines
+# pylint: disable=unused-argument,abstract-method,too-many-lines,useless-suppression
```
[the remark this answers is not in the document]
Comment (maliasadi) [reply frontend/catalyst/jax_primitives.py:80]: hehe.. no it doesn't!

File: frontend/catalyst/jax_tracer.py [hunk 12]
```diff
@@ -435,7 +435,12 @@ def trace_quantum_tape(
     def _bind_native_controlled_op(qrp, op, controlled_wires, controlled_values):
         # For named-controlled operations (e.g. CNOT, CY, CZ) - bind directly by name. For
         # `Controlled(OP)` bind OP with native quantum control syntax.
-        if op.__class__ in {Controlled, ControlledOp, ControlledQubitUnitary}:
+        classinfo = catalyst.api_extensions.quantum_operators.Controlled
```
[the remark this answers is not in the document]
Comment (maliasadi) [reply frontend/catalyst/jax_tracer.py:438]: Ah, I thought I've fixed this.. Good catch!

File: frontend/catalyst/jax_tracer.py (hunk 7 above)
Comment (maliasadi) [reply frontend/catalyst/jax_tracer.py:488]: Not sure I have the context for these meta ops. Could you point me to an example?

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/api_extensions/quantum_operators.py:247]: I see, it would still be a breaking change though, and differs from adjoint now.

File: frontend/catalyst/jax_tracer.py (hunk 7 above)
Comment (dime10) [reply frontend/catalyst/jax_tracer.py:488]: I just mean an operation wrapping other operations in mlir, like what the adjoint is already doing 

```mlir
quantum.adjoint {
  // other ops
}
```

File: frontend/catalyst/api_extensions/quantum_operators.py [hunk 13]
```diff
@@ -656,6 +691,69 @@ def map_wires(self, wire_map):
...
+        # Added this condition to support direct calls of this class outside the QJIT context
+        # in PL integrated tests:
+        if tracing_artifacts:
+            HybridCtrl.__init__(
+                self,
+                *tracing_artifacts,
+                control_wires=control_wires,
+                control_values=control_values,
+                work_wires=work_wires,
+            )
```
[the remark this answers is not in the document]
Comment (maliasadi) [reply frontend/catalyst/api_extensions/quantum_operators.py:725]: Actually I had to use this if-condition to get `Controlled._flatten` PL tests pass when constructing a `Controlled` class outside the context of QJIT; the problem here is that there is not default value for `in_classical_tracers`, `out_classical_tracers`, and `regions`. But you have a good point here and the alternative approach that would fix the issue is 
 
```
            HybridCtrl.__init__(
                self,
                *tracing_artifacts if tracing_artifacts else ([], [], []),
                control_wires=control_wires,
                control_values=control_values,
                work_wires=work_wires,
            )

```
what do you think?

File: frontend/catalyst/jax_tracer.py (hunk 7 above)
Comment (maliasadi) [reply frontend/catalyst/jax_tracer.py:487]: ah, right! gotta :)

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (maliasadi) [reply frontend/catalyst/api_extensions/quantum_operators.py:247]: let me then revert it back to `f`.

File: frontend/test/lit/test_quantum_control.py (hunk 8 above)
Comment (maliasadi) [reply frontend/test/lit/test_quantum_control.py:134]: This is now lowered to QubitUnitary with an updated matrix if it makes sense.

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/api_extensions/quantum_operators.py]: Default values can be produced in `trace_body` no?

File: frontend/test/lit/test_quantum_control.py (hunk 8 above)
Comment (dime10) [reply frontend/test/lit/test_quantum_control.py:134]: Can you point out what effected that change?

Comment (maliasadi) [reply frontend/test/lit/test_quantum_control.py:134]: `qml.ControlledQubitUnitary` is traced and lowered as before,

```
    %4 = stablehlo.convert %arg0 : (tensor<2x2xf64>) -> tensor<2x2xcomplex<f64>>
    %out_qubits, %out_ctrl_qubits:2 = quantum.unitary(%4 : tensor<2x2xcomplex<f64>>) %1 ctrls(%2, %3) ctrlvals(%true, %true) : !quantum.bit ctrls !quantum.bit, !quantum.bit
    %5 = quantum.extract %0[ 3] : !quantum.reg -> !quantum.bit
    %6 = quantum.extract %0[ 4] : !quantum.reg -> !quantum.bit
    %7 = quantum.compbasis %out_ctrl_qubits#0, %out_ctrl_qubits#1, %out_qubits, %5, %6 : !quantum.obs
    %8 = quantum.state %7 : tensor<32xcomplex<f64>>
    %9 = quantum.insert %0[ 2], %out_qubits : !quantum.reg, !quantum.bit
    %10 = quantum.insert %9[ 0], %out_ctrl_qubits#0 : !quantum.reg, !quantum.bit
    %11 = quantum.insert %10[ 1], %out_ctrl_qubits#1 : !quantum.reg, !quantum.bit
```

However, `qml.ctrl(qml.QubitUnitary)` is no longer decomposed to `ControlledQubitUnitrary` and instead it lowers directly to `QubitUnitary` with an updated matix,

```
    %out_qubits:3 = quantum.unitary(%arg0 : tensor<8x8xcomplex<f64>>) %1, %2, %3 : !quantum.bit, !quantum.bit, !quantum.bit
```

There is no concern with the validity of results.

Comment (dime10) [reply frontend/test/lit/test_quantum_control.py:134]: Right you described the change in observable behaviour, but I'd like to understand why that change happened. What part of the PR exactly modified the decomposition behaviour, and why? 

I'm not concerned about the validity here, more so that we fully understand the code we write 😄

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/api_extensions/quantum_operators.py]: > tracing_artifacts if tracing_artifacts else ([], [], [])

This doesn't seem quite right either because that means the HybridOp has no nested tape? Even if a HybridOp is generated outside a qjit context, it may not have input or output tracers but it should still have a nested region 🤔

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 9 above)
Comment (maliasadi) [reply frontend/catalyst/api_extensions/quantum_operators.py:774]: yeah, I tried with `ctrl` but it didn't work..

Comment (maliasadi) [reply frontend/catalyst/api_extensions/quantum_operators.py:774]: But no! I tested before supporting generic PL's `Operator`! Clever idea @dime10 :+1: Let me try it out.

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/api_extensions/quantum_operators.py]: Wait I think your existing implementation already produces  `[], [], [region]` when used outside of a qjit context. Are you sure `tracing_artifacts` can ever be False?

File: frontend/test/lit/test_quantum_control.py (hunk 8 above)
Comment (maliasadi) [reply frontend/test/lit/test_quantum_control.py:134]: Yeah, I know and I'm still investigating :smile:

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (maliasadi) [reply frontend/catalyst/api_extensions/quantum_operators.py]: Ah, there was an issue with the `_unflatten` method. Fixed it! [cf70d1b](https://github.com/PennyLaneAI/catalyst/pull/771/commits/cf70d1b140e7417ff158dc5a43e930b053c17258)

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 9 above)
Comment (maliasadi) [reply frontend/catalyst/api_extensions/quantum_operators.py:788]: Done! [cf70d1b](https://github.com/PennyLaneAI/catalyst/pull/771/commits/cf70d1b140e7417ff158dc5a43e930b053c17258)

File: frontend/catalyst/api_extensions/quantum_operators.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (maliasadi) [reply frontend/catalyst/api_extensions/quantum_operators.py]: Fixed! [cf70d1b](https://github.com/PennyLaneAI/catalyst/pull/771/commits/cf70d1b140e7417ff158dc5a43e930b053c17258)

**Review by dime10 — APPROVED** [review]
🚀

File: frontend/test/lit/test_quantum_control.py (hunk 8 above)
Comment (maliasadi) [reply frontend/test/lit/test_quantum_control.py:134]: The difference here comes from the fact that, previously, `qml.ctrl(qml.QubitUnitary)` was an instance of `QCtrl` and [as this is not natively supported in lightning-qubit](https://github.com/PennyLaneAI/pennylane-lightning/blob/9b933a23fbf6b2177b393dc265ea3dcb6bde541d/pennylane_lightning/lightning_qubit/lightning_qubit.toml#L30), [the PL Controlled decomposition method was converting](https://github.com/PennyLaneAI/pennylane/blob/7f97ea735b3c43b43a0ffe34040bffaa49603268/pennylane/ops/op_math/controlled.py#L252) the gate to  `qml.ControlledQubitUnitary`. Now, we bypass the custom decomposition method of PL's Controlled and fully use Catalyst decomposition rules based on the device capabilities instead. Updating `lightning_qubit.toml` in the Lightning repo will fix this inconsistency:
```
QubitUnitary           = { properties = [ "invertible", "controllable", ] }
```
